hostsfile/src/lib.rs

```diff
@@ -206,9 +209,33 @@ impl HostsBuilder {
         let begin = lines.iter().position(|line| line.trim() == begin_marker);
         let end = lines.iter().position(|line| line.trim() == end_marker);
 
+        let mut lines_to_insert = vec![];
+        if !self.hostname_map.is_empty() {
+            lines_to_insert.push(begin_marker);
+            for (ip, hostnames) in &self.hostname_map {
+                if cfg!(windows) {
+                    // windows only allows one hostname per line
+                    for hostname in hostnames {
+                        lines_to_insert.push(format!("{ip} {hostname}"));
+                    }
+                } else {
+                    // assume the same format as Unix
+                    lines_to_insert.push(format!("{} {}", ip, hostnames.join(" ")));
+                }
+            }
+            lines_to_insert.push(end_marker);
+        }
+
         let insert = match (begin, end) {
             (Some(begin), Some(end)) => {
-                lines.drain(begin..end + 1);
+                let old_section: Vec<String> = lines.drain(begin..end + 1).collect();
+
+                let sections_are_eq = old_section == lines_to_insert;
+
+                if sections_are_eq {
```

Comment:
This should be now clear enough not to introduce and intermediate variable:
```suggestion
                if old_section == lines_to_insert {
```